docarray/index/backends/qdrant.py

```diff
@@ -0,0 +1,318 @@
+from grpc._channel import _InactiveRpcError
+from qdrant_client.http.exceptions import UnexpectedResponse
+
+import docarray.typing.id
+from docarray import BaseDoc, DocArray
+from docarray.index.abstract import BaseDocIndex, _FindResultBatched, _ColumnInfo
+
+import qdrant_client
+from qdrant_client.conversions import common_types as types
+from qdrant_client.http import models as rest
+
+from docarray.typing.tensor.abstract_tensor import AbstractTensor
+from docarray.utils._internal.misc import torch_imported
+from docarray.utils.find import _FindResult
+
+TSchema = TypeVar('TSchema', bound=BaseDoc)
+
+
+QDRANT_PY_VEC_TYPES: List[Any] = [np.ndarray, AbstractTensor]
+if torch_imported:
+    import torch
+    QDRANT_PY_VEC_TYPES.append(torch.Tensor)
+
+QDRANT_SPACE_MAPPING = {
+    'cosine': rest.Distance.COSINE,
+    'l2': rest.Distance.EUCLID,
+    'ip': rest.Distance.DOT,
+}
+
+
+class QdrantDocumentIndex(BaseDocIndex, Generic[TSchema]):
+
+    UUID_NAMESPACE = uuid.UUID('3896d314-1e95-4a3a-b45a-945f9f0b541d')
+
+    def __init__(self, db_config=None, **kwargs):
+        super().__init__(db_config=db_config, **kwargs)
+        self._db_config = cast(QdrantDocumentIndex.DBConfig, self._db_config)
+        self._client = qdrant_client.QdrantClient(
+            url=self._db_config.url,
+            port=self._db_config.port,
+            grpc_port=self._db_config.grpc_port,
+            prefer_grpc=self._db_config.prefer_grpc,
+            https=self._db_config.https,
+            api_key=self._db_config.api_key,
+            prefix=self._db_config.prefix,
+            timeout=self._db_config.timeout,
+            host=self._db_config.host,
+        )
+        self._initialize_collection()
+        self._logger.info(f'{self.__class__.__name__} has been initialized')
+
+    @dataclass
+    class DBConfig(BaseDocIndex.DBConfig):
+        url: Optional[str] = None
+        port: Optional[int] = 6333
+        grpc_port: int = 6334
+        prefer_grpc: bool = True
+        https: Optional[bool] = None
+        api_key: Optional[str] = None
+        prefix: Optional[str] = None
+        timeout: Optional[float] = None
+        host: Optional[str] = None
+        collection_name: str = 'documents'
+        shard_number: Optional[int] = None
+        replication_factor: Optional[int] = None
+        write_consistency_factor: Optional[int] = None
+        on_disk_payload: Optional[bool] = None
+        hnsw_config: Optional[types.HnswConfigDiff] = None
+        optimizers_config: Optional[types.OptimizersConfigDiff] = None
+        wal_config: Optional[types.WalConfigDiff] = None
+        quantization_config: Optional[types.QuantizationConfig] = None
+
+    @dataclass
+    class RuntimeConfig(BaseDocIndex.RuntimeConfig):
+        default_column_config: Dict[Type, Dict[str, Any]] = field(
+            default_factory=lambda: {
+                'id': {},
+                'vector': {},
+                'payload': {},
+                np.ndarray: {},
```

Comment:
Is `np.ndarray` really needed here?

Reply:
I reused the following test from hnswlib:

```
def test_schema_with_user_defined_mapping(tmp_path):
    class MyDoc(BaseDoc):
        tens: NdArray[10] = Field(dim=1000, col_type=np.ndarray)

    store = HnswDocumentIndex[MyDoc](work_dir=str(tmp_path))
    assert store._column_infos['tens'].db_type == np.ndarray
```

When the `col_type` is defined for a field, it's passed directly without calling `python_type_to_db_type`. I wondered how to overcome it but couldn't find a better way. How should I handle that?

Reply:
I believe passing `ndarray` as a column type only really makes sense for hnswlib, not for Qdrant. The user should pass a column type that is a valid type in Qdrant, i.e. one of the ones that you define in your mapping ('vector', 'payload', etc.). No special handling should be necessary

Reply:
Done, the np.array is not here anymore. The three types are left though, as we need to differentiate the behaviour for ids, vectors and metadata attributes (those are stored differently).